### PR TITLE
doc: Honor SOURCE_DATE_EPOCH for documentation timestamps.

### LINKED ISSUE #4262: Honor `SOURCE_DATE_EPOCH` environment variable when generating library documentation
This leads to nondeterminism in the build artifacts (the documentation) of RobotFramework itself or its libraries.  The issue was discovered at the time RobotFramework was packaged for GNU Guix.  For more context and the associated [pull request](https://github.com/robotframework/robotframework/pull/3140).

### PR DESCRIPTION
**Pull request by Apteryks**
Fixes #4262.

* doc/userguide/ug2html.py (create_userguide): Do not embed
timestamps.
* src/robot/utils/robottime.py (get_timestamp_for_doc): New procedure.
* src/robot/utils/__init__.py: Export it.
* src/robot/libdocpkg/model.py (LibraryDoc.to_dictionary): Use it.
* src/robot/libdocpkg/xmlwriter.py (LibdocXmlWriter._write_start):
Likewise.
* BUILD.rst: Document it.

### CODE REVIEW

**Comment by Apteryks** [comment]
OK, so this seems to have broken acceptance tests, due to slightly changing the timestamp format used in XML generated files.  I'll try looking at it later.

File: src/robot/libdocpkg/xmlwriter.py
Comment (pekkaklarck) [inline src/robot/libdocpkg/xmlwriter.py:37]: This changes the timestamp from current `2022-05-27T19:08:42Z` to `2022-05-27T19:07:15+00:00`. That ought to be the same thing, and mach `xs:dateTime` that we use in `libdoc.xsd` schema, but because external tools may depend on the exact format this makes the change backwards incompatible. It's not so big change it couldn't be done in RF 5.1, but need to still be mentioned in the release notes. I'll add a label to the issue so that we don't forget about it.

File: src/robot/libdocpkg/model.py
Comment (pekkaklarck) [inline src/robot/libdocpkg/model.py:116]: This changes the current timestamp from `2022-05-27 22:07:15` to `2022-05-27T19:07:15+00:00`. The old timestamp was naive and I guess it makes sense to have the timezone information there. The change is only in HTML output so it shouldn't cause real backwards compatibility problems. I don't particularly like the `T` separator in outputs targeted for humans, but don't care strongly about it.

File: src/robot/utils/robottime.py
Comment (pekkaklarck) [inline src/robot/utils/robottime.py:327]: This utility is used only in two places and only by Libdoc. I don't think adding it to `robot.utils` is a good idea because it's part of our public API and we cannot change it in the future without backwards compatibility concerns. There's no good place to have utils in the `robot.libdocpkg` package, but I think it would be fine to just inline these few lines of code. It would then make it possible to have slightly different formatting in the timestamp as well.

**Review by pekkaklarck — CHANGES_REQUESTED** [review]
Looks good but I think it's better to inline the timestamp generation code than have it in `robot.utils` that's part of our public API. Change to some tests seem unnecessary as well. See the more detailed comments for details.

File: src/robot/libdocpkg/model.py
Comment (pekkaklarck) [reply src/robot/libdocpkg/model.py:116]: Ooops, my comment was partly invalid. This changes also JSON spec files, not only HTML output. Thus this change is backwards incompatible the same way as the change to XML spec files. Also makes the `T` separator pretty much mandatory.

File: atest/robot/standard_libraries/operating_system/modified_time.robot
Comment (Apteryks) [reply atest/robot/standard_libraries/operating_system/modified_time.robot:11]: No, there're needed because with the ability to set the date all the way back to 1969 (epoch == 0), the date stamp prefix no longer is 20 and the tests would fail.

File: src/robot/utils/robottime.py
Comment (Apteryks) [reply src/robot/utils/robottime.py:327]: Since the timestamp gets output in various files, you can't change it at any time lightly anyway, so I don't see the concern to have the procedure generating it live in a public API place.  As both a maintainer and user, I'd prefer consistency in all dates (figure its format once, and expect it everywhere).

File: BUILD.rst
Comment (Apteryks) [reply BUILD.rst]: I've expound on SOURCE_DATE_EPOCH a bit, thanks!

File: src/robot/utils/robottime.py
Comment (pekkaklarck) [reply src/robot/utils/robottime.py:327]: Actually, there could probably be a `timestamp` property in the `LibraryDoc` model object itself containing this functionality. That would avoid duplication.

**Comment by pekkaklarck** [comment]
One more thing to do is creating tests for this functionality. Need to add tests for JSON and XML specs where SOURCE_DATE_EPOCH is set to some know value before generating specs.

**Review by pekkaklarck — CHANGES_REQUESTED** [review]
Things to do:
- Move the code to generate the timestamp from `robot.utils` to `robot.libdocpkg` to avoid it being part of the public API.
- Revert changes to tests under `atest/{robot,testdata}/standard_libraries` unless they are actually needed. They don't seem to be related to this change and mostly make tests less specific.
- Add test that `SOURCE_DATE_EPOCH` can be used to control the generation time in LIbdoc spec files. Should basically have same test for the generation time as normally but set the environment variable to some know value before running Libdoc. Then we should be able to validate that the exact generation time is what we expect. If you need help with tests, comment here or ping me on Slack.

**Comment by Apteryks** [comment]
> Things to do:
> 
>     * Move the code to generate the timestamp from `robot.utils` to `robot.libdocpkg` to avoid it being part of the public API.

OK.  That seems simple enough.  I'll look into it.

>     * Revert changes to tests under `atest/{robot,testdata}/standard_libraries` unless they are actually needed. They don't seem to be related to this change and mostly make tests less specific.

These changes are actually needed, as I wrote here https://github.com/robotframework/robotframework/pull/4286/files#r883940643.  Other adjustments are needed due to the format having changed slightly.

>     * Add test that `SOURCE_DATE_EPOCH` can be used to control the generation time in LIbdoc spec files. Should basically have same test for the generation time as normally but set the environment variable to some know value before running Libdoc. Then we should be able to validate that the exact generation time is what we expect. If you need help with tests, comment here or ping me on Slack.

Thanks, I'll look into adding an extra test here as well.

File: src/robot/libdocpkg/xmlwriter.py
Comment (pekkaklarck) [inline src/robot/libdocpkg/xmlwriter.py:17]: We've used relative imports with package internal imports. That would change these imports to:

```python
from robot.utils import XmlWritre

from .output import get_timestamp_for_doc
```

File: src/robot/libdocpkg/output.py
Comment (pekkaklarck) [inline src/robot/libdocpkg/output.py:47]: I'd prefer `get_generation_time` as the function name.

File: atest/robot/libdoc/spec_library.robot
Comment (pekkaklarck) [inline atest/robot/libdoc/spec_library.robot:2]: Ought to be already imported by the imported resource file.

File: atest/robot/libdoc/spec_library.robot
Comment (pekkaklarck) [inline atest/robot/libdoc/spec_library.robot:122]: This is _really_ complicated. Why not just set `SOURCE_DATE_EPOCH` to a know value and validate that generation time is expected? I see no point in checking is the environment variable set earlier.

The environment variable should probably then be removed in teardown to avoid subsequent tests being affected. That obviously changes environment of the current process if `SOURCE_DATE_EPOCH` was set in the system earlier, but no acceptance test should rely on external environment variables like that. If considered important, that environment variable could be cleared automatically either by Libdoc tests or even in `atest/run.py` before starting execution.

**Review by pekkaklarck — CHANGES_REQUESTED** [review]
The created test is _really_ complicated and needs to be simplified.

Would also be good to revert unrelated test changes. Or are they actually needed?

I also added some nitpickey comments about imports and function name, but I can change those after the merge myself as well.

File: atest/robot/libdoc/spec_library.robot
Comment (Apteryks) [reply atest/robot/libdoc/spec_library.robot:122]: Clearing the value in atest/run.py should be doable, good idea. I'll try it.

File: atest/robot/libdoc/spec_library.robot
Comment (pekkaklarck) [reply atest/robot/libdoc/spec_library.robot:2]: There's no single best practice but we have generally imported only resource files in our tests if those resource files anyway import libraries we need. One use for resource files with Robot is to avoid importing multiple libraries or resource files in a test case file to keep them simpler.

File: src/robot/libdocpkg/xmlwriter.py
Comment (pekkaklarck) [reply src/robot/libdocpkg/xmlwriter.py:17]: Interesting, I don't remember seeing that recommendation in PEP-8 earlier. Anyway, our style has been doing relative inter-package imports and only use absolute imports when importing something from other top-level packages. I think it makes sense with packages with lot of sub-packages in general, but most importantly imports should use consistent style.

File: atest/robot/standard_libraries/operating_system/modified_time.robot
Comment (pekkaklarck) [reply atest/robot/standard_libraries/operating_system/modified_time.robot:11]: That would require changing system time wouldn't it? Why would you ever want to do that when running Robot's acceptance tests?

File: atest/robot/libdoc/spec_library.robot
Comment (Apteryks) [reply atest/robot/libdoc/spec_library.robot:2]: I see!

File: atest/robot/libdoc/spec_library.robot
Comment (Apteryks) [reply atest/robot/libdoc/spec_library.robot]: I've implemented a much simplified (albeit less clean) version in the latest push.

**Review by pekkaklarck — APPROVED** [review]

**Pull request merged**
